### ISSUE TITLE
`update_repo`: Run 2x to alleviate errors

### ISSUE THREAD

**Issue by joeflack4** [INCATools/ontology-development-kit#989]
## Overview
I'm not sure why, but `update_repo` currently produces errors the first time you run it (after any changes): https://github.com/monarch-initiative/mondo-ingest/issues/308

Until that gets fixed, the current practice is that users must run `update_repo` a second time. I propose that rather than requiring users to know that they need to do this and remember to do it, we automatically run `update_repo` 2x in a row. See: [ignoring errors in make](https://www.gnu.org/software/make/manual/html_node/Errors.html#:~:text=To%20ignore%20errors%20in%20a,to%20the%20shell%20for%20execution.)

**Comment by matentzn** [comment]
I thought about this problem for a bit now, also with ChatGPT, and some braincells active (at least 6 of my 23 braincells). I cannot figure out how to circumvent the problem. The `run.sh` is a wrapper script that calls the `make` command which calls the `update_script.sh` which updates the `run.sh` script - causing the issue. Since I cant do anything outside the `run.sh` command the only solution I would see is to "capture" the `update_repo` command in the `run.sh` script, and wait basically to the last second (before the scripts exits) with updating run.sh. Not even sure if that would work.

Brain teaser!

**Comment by gouttegd** [comment]
Not saying this is the right solution, but a separate launcher like the one evoked [here](https://github.com/INCATools/ontology-development-kit/issues/877) and [here](https://github.com/INCATools/ontology-development-kit/issues/908) would be one way to get rid of this issue, because the launcher would no longer be part of the files that are updated by `update_repo`.

I have a PoC, [experimental binary launcher](https://github.com/gouttegd/odkrunner) for the ODK in the works.

**Comment by matentzn** [comment]
WOW! Awesome! Surely keeping track of _that_!

**Comment by gouttegd** [comment]
Don’t expect fast progress! This is a side-project of a side-project.

**Comment by gouttegd** [comment]
As already noted, using the `odkrun` launcher allows to circumvent the problem, but since we can’t rely on that launcher being available (it’s still a “side project”), I think there is one completely different approach we could try: implementing the `update-repo` logic directly in the seeding script (`odk.py`).

That is, have the `odk.py` script accept a new command `update` that would basically do what the `update_repo.sh` script currently does: seeding a “new” repo in a temporary `target` directory then copy the new files from that directory to the real repo.

Benefits:

* The update code that is executed is always the most up-to-date one. Currently, a repo is updated using the `update_repo.sh` script dating from the previous version of the ODK, which will be lacking any new code added to that script in the new version – one of the reasons we recommend running the update twice.
* The update code is entirely in the Docker image, and is at no risk of being overwritten during the update process – the other reason for running the update twice.

The only file that could be overwritten while we are updating remains the `run.sh` script, but a crude workaround for that could be something like:

```sh
# At the very beginning of the script…
# Are we running the update command?
if [ "$1" = update ]; then
    # then make a copy of that script (without this block)...
    sed -n '/^set -e/,$p' $0 > $0.tmp
    # ... end execute the copy
    exec $SHELL $0.tmp "$@"
fi
```

This way it won’t matter that `run.sh` is overwritten during the update, because the shell is running the code in `run.sh.tmp`.

**Comment by joeflack4** [comment]
Detailed as always, Damien.

I do find Python implementations to be more stable than shell script.

FYI I haven't needed to run it more than once for a while, since around when this ticket was opened.

**Comment by gouttegd** [comment]
> I do find Python implementations to be more stable than shell script.

Yes, and in this case doing the update in Python code could also allow us to be more efficient.¹

Because instead of copying all template files in a temporary directory, and then doing a `rsync --ignore-existing` between the temporary directory and the actual repository (which is basically what `update_repo.sh` is doing), we could directly copy/instantiate only the template files that we _know_ need to be updated.

---
¹“Python” and “efficient” in the same sentence? 🤯 Does… not… compute… does… not… compute
